components/brave_new_tab/resources/components/settings/background_panel.tsx

```diff
@@ -0,0 +1,193 @@
+
+import { BackgroundType } from '../../models/new_tab_model'
+import { useNewTabModel, useNewTabState } from '../new_tab_context'
+import { useLocale } from '../locale_context'
+import { useCallbackWrapper } from '../../lib/use_callback_wrapper'
+import { inlineCSSVars } from '../../lib/inline_css_vars'
+import { BackgroundTypePanel } from './background_type_panel'
+
+import {
+  backgroundCSSValue,
+  gradientPreviewBackground,
+  solidPreviewBackground } from '../../models/backgrounds'
+
+import { style } from './background_panel.style'
+
+export function BackgroundPanel() {
+  const { getString } = useLocale()
+  const model = useNewTabModel()
+  const wrapCallback = useCallbackWrapper()
+
+  const [
+    backgroundsEnabled,
+    sponsoredImagesEnabled,
+    selectedBackgroundType,
+    selectedBackground,
+    braveBackgrounds,
+    customBackgrounds
+  ] = useNewTabState((state) => [
+    state.backgroundsEnabled,
+    state.sponsoredImagesEnabled,
+    state.selectedBackgroundType,
+    state.selectedBackground,
+    state.braveBackgrounds,
+    state.customBackgrounds
+  ])
+
+  const [panelType, setPanelType] = React.useState<BackgroundType>('none')
+  const [uploading, setUploading] = React.useState(false)
+
+  function getTypePreviewValue(type: BackgroundType) {
+    const isSelectedType = type === selectedBackgroundType
+    switch (type) {
+      case 'brave':
+        return braveBackgrounds.at(0)?.imageUrl || ''
+      case 'custom':
+        if (isSelectedType && selectedBackground) {
+          return selectedBackground
+        }
+        return customBackgrounds.at(0) ?? ''
+      case 'solid':
+        if (isSelectedType && selectedBackground) {
+          return selectedBackground
+        }
+        return solidPreviewBackground
+      case 'gradient':
+        if (isSelectedType && selectedBackground) {
+          return selectedBackground
+        }
+        return gradientPreviewBackground
+      case 'none':
+        return ''
+    }
+  }
+
+  function renderUploadPreview() {
+    return (
+      <div className='preview upload'>
+        {uploading ? <ProgressRing /> : <Icon name='upload' />}
+        {getString('uploadBackgroundLabel')}
+      </div>
+    )
+  }
+
+  function renderTypePreview(type: BackgroundType) {
+    if (type === 'custom' && customBackgrounds.length === 0) {
+      return renderUploadPreview()
+    }
+    return (
+      <div
+        className='preview'
+        style={inlineCSSVars({
+          '--preview-background':
+            backgroundCSSValue(type, getTypePreviewValue(type))
+        })}
+      >
+        {
+          type === selectedBackgroundType &&
+            <span className='selected-marker'>
+              <Icon name='check-normal' />
+            </span>
+        }
+      </div>
+    )
+  }
+
+  function showCustomBackgroundChooser() {
+    model.showCustomBackgroundChooser().then(wrapCallback((selected) => {
+      if (selected && !uploading) {
```

Comment:
won't this just silently fail if it's currently uploading?

Reply:
The UI should not allow the user to get into a state where they are trying to upload an image while an image is currently being uploaded (it will show a spinner instead of the upload button). But if for some reason it did happen, we want to avoid having two uploads in progress and further confusing state.

Reply:
:+1: to this - seems like a pretty reasonable approach